### PR TITLE
*: refine all unit metrics

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 
- None

### What is changed and how it works?

I think all metrics should be treated like in `dm/master/metrics.go`
so, i make some wrapper/setter for unit metrics.

### Check List 

Tests 

- Unit test
- Integration test
- Self Test

Code changes

- [x] `dm/worker/metrics.go`
- [x] `relay/metrics.go`
- [x] `loader/metrics.go`
- [x] `dumpling/metrics.go`
- [x] `syncer/metrics.go`

Side effects

 - **Increased code complexity**
 
Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Review by lichunzhu — COMMENTED** [review]
What's the benefit of wrapping these vectors?

**Comment by Ehco1996** [comment]
> What's the benefit of wrapping these vectors?

* all metrics in dm-master is wrapped, so I think we should uniform the code？
* Provide the exported func in case other package need use it

File: dm/worker/metrics.go [hunk 1]
```diff
@@ -138,3 +139,17 @@ func InitStatus(lis net.Listener) {
 		log.L().Error("status server returned", log.ShortError(err))
 	}
 }
+
+// UpdatetaskStateGauge update worker task stage.
+func UpdatetaskStateGauge(task, sourceID string, stage pb.Stage) {
```
Comment (lance6716) [inline dm/worker/metrics.go:144]:
```suggestion
func UpdateTaskStateGauge(task, sourceID string, stage pb.Stage) {
```

Comment (lance6716) [reply dm/worker/metrics.go:144]: And I'm not sure if we should delete the `Stage_Finished` tasks. Now I prefer don't delete them if they can be seen from `query-status` 😂

File: dumpling/dumpling.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func (m *Dumpling) Init(ctx context.Context) error {
 
 // Process implements Unit.Process.
 func (m *Dumpling) Process(ctx context.Context, pr chan pb.ProcessResult) {
-	dumplingExitWithErrorCounter.WithLabelValues(m.cfg.Name, m.cfg.SourceID).Add(0)
+	AddDumplingExitWithErrorCounter(float64(0), m.cfg.Name, m.cfg.SourceID)
```
Comment (lance6716) [inline dumpling/dumpling.go:69]:
```suggestion
	AddDumplingExitWithErrorCounter(0, m.cfg.Name, m.cfg.SourceID)
```

**Review by lance6716 — COMMENTED** [review]
will review later

File: dm/worker/metrics.go (hunk 1 above)
Comment (Ehco1996) [reply dm/worker/metrics.go:144]: i vote for keep stage  `Stage_Finished`😂

File: pkg/binlog/event/util.go [hunk 3]
```diff
@@ -218,7 +218,7 @@ func encodeColumnValue(v interface{}, tp byte, meta uint16) ([]byte, error) {
 	case gmysql.MYSQL_TYPE_DOUBLE:
 		value, ok := v.(float64)
 		if !ok {
-			err = terror.ErrBinlogColumnTypeMisMatch.Generate(v, reflect.TypeOf(v), reflect.TypeOf(float64(0)))
+			err = terror.ErrBinlogColumnTypeMisMatch.Generate(v, reflect.TypeOf(v), reflect.TypeOf(0))
```
Comment (lance6716) [inline pkg/binlog/event/util.go:221]: should keep float64(0) here, float64(0) and 0 has different type

File: syncer/metrics.go [hunk 4]
```diff
@@ -273,23 +273,156 @@ func InitStatusAndMetrics(addr string) {
...
+func AddJobCountMetric(s *Syncer, isFinished bool, queueBucket string, tp opType, n int64) {
+	m := addedJobsTotalCounter
+	if isFinished {
+		s.count.Add(n)
+		m = finishedJobsTotalCounter
+	}
+
+	switch tp {
+	case insert:
+		m.WithLabelValues("insert", s.cfg.Name, queueBucket, s.cfg.SourceID).Add(float64(n))
```
Comment (lance6716) [inline syncer/metrics.go:412]: "insert" could be retrieved by tp.String()

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: syncer/metrics.go (hunk 4 above)
Comment (Ehco1996) [reply syncer/metrics.go]: added

**Comment by Ehco1996** [comment]
close for now

**Pull request closed without merging**